ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1493 @@
+---
+title: Workflow-Engine Stream Processing 
+authors:
+  - Philipp Ossler (@saig0)
+  - Nico Korthout (@korthout)
+reviewers:
+  - Peter Ihme (@pihme)
+  - Nico Korthout (@korthout)
+approvers:
+  - TBD
+editor: NA
+creation-date: 2020-11-02
+last-updated: 2020-12-21
+status: provisional
+see-also: 
+  - "Build state on followers" 
+replaces: []
+superseded-by: []
+---
+
+# Summary
+[summary]: #summary
+
+Define the way how the workflow engine processes the stream of records, especially how the engine rebuilds its state after a restart.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, we are not able to process the records of a previous Zeebe version if some logic in the workflow engine has changed. The details are described in the issue: https://github.com/zeebe-io/zeebe/issues/5581.
+
+The behavior of the workflow engine when processing a record (event or command):
+* read the state of the workflow instance and related resources (RocksDB)
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state, including the key generator
+* write follow-up events/commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+When the broker is restarted then it first recovers its state from the latest
+snapshot (i.e. RocksDB checkpoint). Afterwards, it reprocesses the records on
+the stream that have already been processed before to rebuild its state. When
+all records are reprocessed then it continues with the normal processing.
+
+The behavior of the workflow engine when reprocessing a record:
+* read the state of the workflow instance and related resources (RocksDB)
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state, including the value/offset of the key generator 
+
+The problem occurs when the business logic of the workflow engine changes (e.g. by fixing a bug, by implementing a new feature, etc.) and it writes different updates to the state on reprocessing. When the next record (of the workflow instance) is reprocessed then the state is different as it was before on processing. This may lead to a corrupted state, or the record can't be reprocessed successfully (e.g. it is rejected by the concurrent update guard, the state doesn't contain the expected data, etc.).  
```

Comment:
This happens also when an bug was fixed and on reprocessing no error record is produced anymore. (Blacklisting again)

Reply:
True. I think we don't need to add more to understand the motivation :)